openedx/core/djangoapps/xblock/README.rst

```diff
@@ -0,0 +1,94 @@
+
+Learning Contexts
+-----------------
+
+A "Learning Context" is a course, a library, a program, or some other collection of content where learning happens.
+
+To work with this runtime, all learnable XBlock content for a learning context must ultimately be stored in a Blockstore bundle, but the runtime does not impose any restrictions beyond that. The relationship between a learning context and a bundle will usually be 1:1, but some future types of learning context may not stick to that, and the runtime doesn't require it.
+
+Currently only `the "Content Library" learning context <./learning_context/content_library/>`_ is implemented. Additional learning contexts can be implemented as plugins, by subclassing the ``LearningContextImpl`` class and registering in the ``openedx.learning_context`` entry point.
+
+A learning context is responsible for:
+
+* Definining whether or not a given XBlock [usage key] exists in the learning context (e.g. "Does the content library X contain block Y?")
+
+  - This is done via its ``definition_for_usage()`` method which must return ``None`` if the usage key doesn't exist in the context.
+  - There is **no** general method to "list all XBlocks" in the learning context, because it's expected that learning contexts may be dynamic, e.g. with content assigned just in time via adaptive learning. However, if a learning context is static it can certainly implement an API to list all the blocks it contains.
+* Determining whether or not a given user has **permission to view/use** a given XBlock and/or to **edit** that XBlock.
+
+  - This is done by via its ``can_view_block()`` and ``can_edit_block()`` methods.
+  - For example, "pathways" might allow any user to view any XBlock, but "courses" require enrollment, cohort, and due date checks as part of this permissions logic.
+
+* Mapping usage keys to ``BundleDefinitionLocator`` keys.
+
+  - The XBlock runtime and other parts of the system do not know nor prescribe how a usage locator like ``lb:library15:html:introduction`` (HTML block with usage ID "introduction" in library with slug "library15") maps to an OLX file in Blockstore like "``html/introduction/definition.xml` in bundle with UUID `11111111-1111-1111-1111-111111111111`" - that logic is left to the learning context.
+  - This is done via its ``definition_for_usage()`` method.
+
+* Definining field overrides: Learning contexts may optionally implement some mechanism for overriding field data found in the Blockstore definitions based on arbitrary criteria.
+
+  - For example, a course may specify a list of XBlock field override rules, such as:
+
+    + "All ``problem`` XBlocks in this course override the ``num_attempts`` field to have a value of ``5``" or
+    + "Users in the ``class B`` group have the ``due_date`` field of all XBlocks adjusted by ``+2 weeks``"
```

Comment:
I was kind of hoping we could skip the overrides bit, but that's a separate conversation. If we are having it, I agree that it makes sense to define at that level.

Reply:
Well I haven't field overrides it nor am I planning to implement them anytime soon, but I have left some placeholder comments throughout the code for where I think overrides would go if we want them. In any case I definitely prefer this "single list of rules" to the existing inheritance model.